### PR TITLE
Make DQC_US_RULE 0041

### PR DESCRIPTION
**Pull request by sagesmith-wf**
#### Changes:

- Created dqc_us_0041.py
- Created unit tests for dqc_us_0041.py
- Created a resource directory for dqc_us_0041's 
- Added cached json files to the cache

Please review: @hefischer  @andrewperkins-wf

### CODE REVIEW

File: dqc_us_rules/dqc_us_0033_0036.py [hunk 1]
```diff
@@ -201,7 +201,7 @@ def _get_default_dped(dped_facts):
         return None
     elif len(keys) == 1:
         # Only one Document Period End Date, so that is our default
-        value, = dped_facts.values()
+        value = dped_facts.values()
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0033_0036.py]: why was this changed?  The ```,``` made it so that the list dereferenced.

A better fix might be to do ```value = dped_facts.values()[0]```

File: dqc_us_rules/dqc_us_0041.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
+# (c) Copyright 2015 - 2016, XBRL US Inc. All rights reserved.
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: Copyright should just be 2016.

File: dqc_us_rules/dqc_us_0041.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+):
+    """
+    Creates a new caches for the Taxonomy default dimensions
+
+    :param val: ValidateXbrl to be validated
+    :type val: :class: '~arelle.ValidateXbrl.ValidateXbrl'
+    :param ugt: Taxonomy to check
+    :type ugt: str
+    :param cntlr: cntlr to save to
+    :type cntlr: :class: '~arelle.CntrlWinMain'
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: The type here should be ```:~arelle.Cntlr.Cntlr:```.  It won't be a ```CntlrWinMain``` except when running with the GUI.

File: dqc_us_rules/dqc_us_0041.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+    :param ugt_calcs_json_file: where to save json taxonomies
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: This would be where to save the calcs json cache, right?  And do we need to be saving the calculations cache?

File: dqc_us_rules/dqc_us_0041.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+    :type ugt_calcs_json_file: str
+    :param ugt_default_dimensions_json_file: location to save json default
+        dimensions
+    :type ugt_default_dimensions_json_file: str
+    :return: no explicit return, but saves caches for dqc_us_0041
+    :rtype: None
+    """
+    val.modelXbrl.modelManager.addToLog(
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: We should not be logging informational messages this way.  If we want to log informational messages, we should be using a logger created at our class level.

File: dqc_us_rules/resources/DQC_US_0041/ugt-default-dimensions.json [hunk 6]
```diff
@@ -0,0 +1,273 @@
+{
```
Comment (andrewperkins-wf) [inline dqc_us_rules/resources/DQC_US_0041/ugt-default-dimensions.json]: What year is this for?

File: dqc_us_rules/dqc_us_0041.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+                    _("loading us-gaap {0} calculations and default "  # noqa
+                      "dimensions into cache").format(ugt["year"])
+                )
+    started_at = time.time()
+    ugt_entry_xsd = ugt["entryXsd"]
+    val.usgaapCalculations = {}
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: we shouldn't be saving the calculations at this time.  We are not using them that I can see.

File: dqc_us_rules/dqc_us_0041.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    val.usgaapDefaultDimensions = {}
+    prior_validate_disclosure_system = (
+        val.modelXbrl.modelManager.validateDisclosureSystem
+    )
+    val.modelXbrl.modelManager.validateDisclosureSystem = False
+    calculations_instance = (ModelXbrl.load(
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: Another calculation piece

File: dqc_us_rules/dqc_us_0041.py [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+        val.modelXbrl.modelManager,
+        openFileSource(ugt_entry_xsd, cntlr),
+        _("built us-gaap calculations cache"))  # noqa
+    )
+    val.modelXbrl.modelManager.validateDisclosureSystem = (
+        prior_validate_disclosure_system
+    )
+    if calculations_instance is None:
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: more calculations

File: dqc_us_rules/dqc_us_0041.py [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+        val.modelXbrl.error(
+            "arelle:notLoaded",
+            _("US-GAAP calculations not loaded: %(file)s"),  # noqa
+            modelXbrl=val,
+            file=os.path.basename(ugt_entry_xsd)
+        )
+    else:
+        # load calculations
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: it looks like this whole ```else``` block is only for calculations

**Comment by andrewperkins-wf** [comment]
@sagesmith-wf First pass is complete - remove the calculations stuff, and that should clean it up a LOT for us to do a more in depth review.

File: dqc_us_rules/dqc_us_0041.py [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+    """
+    file = None
+    try:
+        file = openFileStream(
+            cntlr,
+            ugt_calcs_json_file,
+            'rt',
+            encoding='utf-8'
+        )
+        val.usgaapCalculations = json.load(file)
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: I don't think we need to load the calc cached data.

File: dqc_us_rules/dqc_us_0041.py [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+import re
+import time
+
+from arelle.ModelDtsObject import ModelConcept
+from arelle import XbrlConst, ModelXbrl
+from arelle.FileSource import openFileStream, saveFile, openFileSource
+
+
+_CODE_NAME = 'DQC.US.0041'
+_RULE_VERSION = '1.0'
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: This should probably be ```1.1``` since it is being added in the 1.1 release.

File: dqc_us_rules/dqc_us_0041.py (hunk 8 above)
Comment (sagesmith-wf) [inline dqc_us_rules/dqc_us_0041.py]: I need this to get the default dimensions, but I don't have to save it so I have removed that part

File: dqc_us_rules/resources/DQC_US_0041/ugt-default-dimensions.json (hunk 6 above)
Comment (sagesmith-wf) [inline dqc_us_rules/resources/DQC_US_0041/ugt-default-dimensions.json]: This is currently loading the first one that works which is 2015

File: dqc_us_rules/dqc_us_0033_0036.py [hunk 13]
```diff
@@ -31,6 +31,21 @@ def doc_period_end_date_check(val):
         )
 
 
+def _is_valid_eop_fact(eop_fact):
+    """
+    Checks to ensure that the eop_fact and the eop_fact.xValue is not None
+    :param eop_fact: fact to check
```
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0033_0036.py]: SG should have newline between comment and parameter sections.

File: dqc_us_rules/dqc_us_0033_0036.py [hunk 14]
```diff
@@ -31,6 +31,21 @@ def doc_period_end_date_check(val):
...
+    :type eop_fact: :class: "~arelle.InstanceModelObject.ModelFact"
+    :return: True if the fact and the fact.xValue is not None
+    :rtype: bool
+    """
+    if eop_fact is None:
```
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0033_0036.py]: with order of operations and evaluation you can simplify this guy:
```
return False if eop_fact is None or eop_fact.xValue is None else True
```

File: dqc_us_rules/dqc_us_0041.py [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+_CODE_NAME = 'DQC.US.0041'
+_RULE_VERSION = '1.1'
+
+
+ugtDocs = (
+    {"year": 2014,
+     "namespace": "http://fasb.org/us-gaap/2014-01-31",
+     "docLB": "http://xbrl.fasb.org/us-gaap/2014/elts/us-gaap-doc-2014-01-31.xml",  # noqa
+     "entryXsd": "http://xbrl.fasb.org/us-gaap/2014/entire/us-gaap-entryPoint-std-2014-01-31.xsd",  # noqa
+     },
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: SG: Multiline alignment, recommend: 
```
    {
        "year": 2014,
        "namespace": "http://fasb.org/us-gaap/2014-01-31",
        "docLB": "http://xbrl.fasb.org/us-gaap/2014/elts/us-gaap-doc-2014-01-31.xml",  # noqa
        "entryXsd": "http://xbrl.fasb.org/us-gaap/2014/entire/us-gaap-entryPoint-std-2014-01-31.xsd",  # noqa
    },
```
And replicating that down for the other entries

File: dqc_us_rules/dqc_us_0041.py [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+     "namespace": "http://fasb.org/us-gaap/2016-01-31",
+     "docLB": "http://xbrl.fasb.org/us-gaap/2016/us-gaap-2016-01-31.zip/us-gaap-2016-01-31/elts/us-gaap-doc-2016-01-31.xml",  # noqa
+     "entryXsd": "http://xbrl.fasb.org/us-gaap/2016/us-gaap-2016-01-31.zip/us-gaap-2016-01-31/entire/us-gaap-entryPoint-std-2016-01-31.xsd",  # noqa
+     }
+)
+
+
+def _make_cache(
+        val, ugt, cntlr, ugt_default_dimensions_json_file
+):
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: SG: Multiline alignment of a function/class definition, suggest:
```
def _make_cache(val, ugt, cntlr, 
                ugt_default_dimensions_json_file):
```

File: dqc_us_rules/dqc_us_0041.py [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+    val.usgaapDefaultDimensions = {}
+    prior_validate_disclosure_system = (
+        val.modelXbrl.modelManager.validateDisclosureSystem
+    )
+    val.modelXbrl.modelManager.validateDisclosureSystem = False
+    calculations_instance = (ModelXbrl.load(
+        val.modelXbrl.modelManager,
+        openFileSource(ugt_entry_xsd, cntlr),
+        _("built us-gaap calculations cache"))  # noqa
+    )
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: SG: Multiline alignment, recommend: 
```
calculations_instance = (
    ModelXbrl.load(
        val.modelXbrl.modelManager,
        openFileSource(ugt_entry_xsd, cntlr),
        _("built us-gaap calculations cache")  # noqa
    )
)
```

File: dqc_us_rules/dqc_us_0041.py (hunk 16 above)
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0041.py]: actually doesn't look to be >79, so I suggest:
```
def _make_cache(val, ugt, cntlr, ugt_default_dimensions_json_file):
```

File: dqc_us_rules/dqc_us_0041.py [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+        val.modelXbrl.error(
+            "arelle:notLoaded",
+            _("US-GAAP calculations not loaded: %(file)s"),  # noqa
+           modelXbrl=val,
+            file=os.path.basename(ugt_entry_xsd)
+        )
+
+    else:
+        for defaultDimRel in calculations_instance.relationshipSet(
+                XbrlConst.dimensionDefault).modelRelationships:
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: CR: This long ass component should be pulled out into a variable:
```
model_relationships = (
    calculations_instance.relationshipSet(
        XbrlConst.dimensionDefault
    ).modelRelationships
)
for defaultDimRel in model_relationships:
```

File: dqc_us_rules/dqc_us_0041.py [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+            if isinstance(
+                    defaultDimRel.fromModelObject,
+                    ModelConcept
+            ) and isinstance(
+                defaultDimRel.toModelObject,
+                ModelConcept
+            ):
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: CR: This should be extracted into a helper function.

File: dqc_us_rules/dqc_us_0041.py [hunk 20]
```diff
@@ -0,0 +1,294 @@
...
+                from_name = defaultDimRel.fromModelObject.name
+                to_name = defaultDimRel.toModelObject.name
+                val.usgaapDefaultDimensions[from_name] = to_name
+        json_str = str(json.dumps(
+            val.usgaapDefaultDimensions,
+            ensure_ascii=False,
+            indent=0)
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: SG: Multiline alignment, recommend: 
```
json_str = str(
    json.dumps(
        val.usgaapDefaultDimensions,
        ensure_ascii=False,
        indent=0
    )
)
```

File: dqc_us_rules/dqc_us_0041.py [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+            'DQC_US_0041',
+            "_".join([str(year), "ugt-default-dimensions.json"])
+        )
+
+        _make_cache(
+            val,
+            ugt,
+            cntlr,
+            ugt_default_dimensions_json_file
+        )
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: CR: This is kind of tower-like, any reason to not do:
```
_make_cache(
    val, ugt, cntlr,
    ugt_default_dimensions_json_file
)
```

File: dqc_us_rules/dqc_us_0041.py [hunk 22]
```diff
@@ -0,0 +1,294 @@
...
+    :param val: ValidateXbrl that is being validated
+    :type val: :class:'~arelle.ValidateXbrl.ValidateXbrl
+    :param ugt_namespace: Namespace of taxonomy for current year
+    :type ugt_namespace: str
+    :return: True if ugt_namespace is in the val.modelXbrl's namespace docs
+    :rtype: bool
+    """
+    return (ugt_namespace in val.modelXbrl.namespaceDocs and
+            len(val.modelXbrl.namespaceDocs[ugt_namespace]) > 0
+            )
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: SG: Multiline alignment, recommend: 
```
return (
    ugt_namespace in val.modelXbrl.namespaceDocs and
    len(val.modelXbrl.namespaceDocs[ugt_namespace]) > 0
)
```

File: dqc_us_rules/dqc_us_0041.py [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+
+            file = None
+
+            try:
+                file = openFileStream(
+                       cntlr,
+                       ugt_default_dimensions_json_file,
+                       'rt',
+                       encoding='utf-8'
+                )
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: CR: This is also a bit tower like, suggest:
```
file = openFileStream(
    cntlr, ugt_default_dimensions_json_file,
    'rt', encoding='utf-8'
)
```

File: dqc_us_rules/dqc_us_0041.py [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+    rel_set = val.modelXbrl.relationshipSet(
+        XbrlConst.dimensionDefault
+    ).modelRelationships
+    for rel in rel_set:
+        rel_to = rel.toModelObject
+        rel_from = rel.fromModelObject
+        if (_default_dimension_mismatch(
+                rel_to.name,
+                val.usgaapDefaultDimensions[rel_from.name]
+        )):
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: SG: Multiline conditional formatting ([ref](https://github.com/Workiva/styleguide/blob/master/PYTHON.rst#multi-line-ifelifwith-statements)), suggestion:
```
if _default_dimension_mismatch(
    rel_to.name,
    val.usgaapDefaultDimensions[rel_from.name]
):
```
This avoids the ambiguous indent problem, hence the only one set of `( )`

File: dqc_us_rules/dqc_us_0041.py [hunk 25]
```diff
@@ -0,0 +1,294 @@
...
+            yield (rel_from.name,
+                   val.usgaapDefaultDimensions[rel_from.name],
+                   rel_to.name
+                   )
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: SG: Multiline alignment, recommend: 
```
yield (
    rel_from.name,
    val.usgaapDefaultDimensions[rel_from.name],
    rel_to.name
)
```

File: dqc_us_rules/dqc_us_0041.py [hunk 26]
```diff
@@ -0,0 +1,294 @@
...
+
+__pluginInfo__ = {
+    'name': _CODE_NAME,
+    'version': _RULE_VERSION,
+    'description': 'All axis defaults should be the same as the axis '
+                   'defaults defined in the taxonomy.',
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: SG: Multiline alignment, recommend: 
```
'description': (
    'All axis defaults should be the same as the axis '
    'defaults defined in the taxonomy.'
),
```

File: dqc_us_rules/dqc_us_0041.py [hunk 27]
```diff
@@ -0,0 +1,294 @@
...
+    val.ugtNamespace = None
+    cntlr = val.modelXbrl.modelManager.cntlr
+
+    year = 2014
+    for ugt in ugtDocs:
+        ugt_default_dimensions_json_file = os.path.join(
+            os.path.dirname(__file__),
+            'resources',
+            'DQC_US_0041',
+            "_".join([str(year), "ugt-default-dimensions.json"])
```
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0041.py]: typically don't use join unless it's actually joining a list together with a spacer... suggest a format like this:
```
"{}_ugt-default-dimensions.json".format(str(year))
```

File: tests/unit_tests/test_dqc_us_0041.py [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+
+class TestDefaultDimensions(unittest.TestCase):
+    def test_default_dimension_mismatch(self):
+        """
+        Tests to see if _default_dimension_mismatch can determine if two
+        dimensions match or not
+        """
+        bad_domain_pair = ["DebtInstrumentNameDomain", "DebtInstrumentDomain"]
+
+        self.assertTrue(dqc_us_0041._default_dimension_mismatch(
```
Comment (chrislococo-wf) [inline tests/unit_tests/test_dqc_us_0041.py]: SG: Non trivial double context open, please expand to multiple lines.
```
self.assertTrue(
    dqc_us_0041._default_dimension_mismatch(
        bad_domain_pair[0], bad_domain_pair[1]
    )
)
```

File: tests/unit_tests/test_dqc_us_0041.py [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+        """
+        Tests to make sure the _default_dimension_mismatch doesn't always throw
+        an error
+        """
+        good_domain_pair = ["DebtInstrumentDomain", "DebtInstrumentDomain"]
+
+        self.assertFalse(dqc_us_0041._default_dimension_mismatch(
+            good_domain_pair[0],
+            good_domain_pair[1]
+        ))
```
Comment (chrislococo-wf) [inline tests/unit_tests/test_dqc_us_0041.py]: SG: Non trivial double context open, please expand to multiple lines.
suggest:
```
self.assertFalse(
    dqc_us_0041._default_dimension_mismatch(
        good_domain_pair[0], good_domain_pair[1]
    )
)
```

File: dqc_us_rules/dqc_us_0041.py [hunk 30]
```diff
@@ -0,0 +1,294 @@
...
+        )
+
+        _make_cache(
+            val,
+            ugt,
+            cntlr,
+            ugt_default_dimensions_json_file
+        )
+
+        year += 1
```
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0041.py]: I'm not sure I understand what this line is doing... I'm leaning toward nothing...

File: tests/unit_tests/test_dqc_us_0041.py [hunk 31]
```diff
@@ -0,0 +1,80 @@
...
+        val.usgaapDefaultDimensions = usgaap_default_dims
+
+        for _ in dqc_us_0041._catch_dqc_us_0041_errors(val):
+            bad_count += 1
+
+        self.assertTrue(bad_count == 1)
+
+    @patch("dqc_us_rules.dqc_us_0041._load_cache",
+           autospec=True
+           )
```
Comment (chrislococo-wf) [inline tests/unit_tests/test_dqc_us_0041.py]: SG: De-dent to align with `@`

File: dqc_us_rules/dqc_us_0041.py (hunk 16 above)
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: Even better

File: dqc_us_rules/dqc_us_0041.py [hunk 32]
```diff
@@ -0,0 +1,294 @@
...
+    year = 2014
+
+    for ugt in ugtDocs:
+        ugt_namespace = ugt["namespace"]
+        if _is_in_namespace(val, ugt_namespace):
+            ugt_default_dimensions_json_file = os.path.join(
+                os.path.dirname(__file__),
+                'resources',
+                'DQC_US_0041',
+                "_".join([str(year), "ugt-default-dimensions.json"])
```
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0041.py]: again, should use .format here instead of .join

File: dqc_us_rules/dqc_us_0041.py (hunk 27 above)
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: ^ I like that, that is cleaner.

File: dqc_us_rules/dqc_us_0041.py (hunk 30 above)
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0041.py]: nevermind, I didn't realize this was in a loop... maybe consider having years supported in a list at top, and referencing their index by your incrementing through the loop? just a suggestion

File: dqc_us_rules/dqc_us_0041.py [hunk 33]
```diff
@@ -0,0 +1,294 @@
...
+                       encoding='utf-8'
+                )
+
+                val.usgaapDefaultDimensions = json.load(file)
+                file.close()
+
+            except FileNotFoundError:
+                if file:
+                    file.close()
+        year += 1
```
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0041.py]: but if you do take the suggestion for having a list of supported years, be sure to use it here too

**Comment by brianmyers-wf** [comment]
pass complete

File: dqc_us_rules/dqc_us_0041.py (hunk 33 above)
Comment (sagesmith-wf) [inline dqc_us_rules/dqc_us_0041.py]: Extracted start year out to constant, but since I am already iterating over list so I couldn't take your suggestion

File: dqc_us_rules/dqc_us_0041.py (hunk 30 above)
Comment (sagesmith-wf) [inline dqc_us_rules/dqc_us_0041.py]: I am incrementing the year here for my file name

**Comment by chrislococo-wf** [comment]
+1

File: dqc_us_rules/dqc_us_0041.py [hunk 34]
```diff
@@ -16,31 +16,33 @@
 _CODE_NAME = 'DQC.US.0041'
 _RULE_VERSION = '1.1'
 
+_EARLIEST_GAP_YEAR = 2014
```
Comment (brianmyers-wf) [inline dqc_us_rules/dqc_us_0041.py]: Hate to bring this up, but it should be
```
_EARLIEST_GAAP_YEAR
```
notice the two a's

**Comment by brianmyers-wf** [comment]
+1

**Comment by chrislococo-wf** [comment]
+1

File: dqc_us_rules/dqc_us_0041.py [hunk 35]
```diff
@@ -0,0 +1,312 @@
...
+    :rtype: None
+    """
+    started_at = time.time()
+    ugt_entry_xsd = ugt["entryXsd"]
+    val.usgaapDefaultDimensions = {}
+    prior_validate_disclosure_system = (
+        val.modelXbrl.modelManager.validateDisclosureSystem
+    )
+    val.modelXbrl.modelManager.validateDisclosureSystem = False
+    calculations_instance = (
```
Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: This should probably be called something else, right?  Since we aren't doing calcs with it anymore

**Comment by andrewperkins-wf** [comment]
+1 as long as my comment on naming is handled in a followup.

File: dqc_us_rules/dqc_us_0041.py (hunk 35 above)
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: Followup ticket created here: https://jira.atl.workiva.net/browse/XS-2999

File: dqc_us_rules/dqc_us_0041.py [hunk 36]
```diff
@@ -288,17 +304,19 @@ def _catch_dqc_us_0041_errors(val):
...
-        if _default_dimension_mismatch(
-            rel_to.name,
-            val.usgaapDefaultDimensions[rel_from.name]
-        ):
-            yield (
-                rel_from.name,
-                val.usgaapDefaultDimensions[rel_from.name],
-                rel_to.name
-            )
+        if _check_relationship_exists(rel):
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: CR: @andrewperkins-wf should nothing be thrown if the relationship does not exist, or does that still violate the spirit of the rule?

File: dqc_us_rules/dqc_us_0041.py [hunk 37]
```diff
@@ -0,0 +1,330 @@
...
+        )
+
+    else:
+        model_relationships = (
+            ugt_entry_xsd_instance.relationshipSet(
+                XbrlConst.dimensionDefault
+            ).modelRelationships
+        )
+        for default_dim_rel in model_relationships:
+            if _default_dim_rel_is_instance(default_dim_rel):
```
Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: CR: @andrewperkins-wf should nothing be thrown if the relationship does not exist, or does that still violate the spirit of the rule?

Comment (andrewperkins-wf) [inline dqc_us_rules/dqc_us_0041.py]: If nothing exists for that relationship, they don't have a default member, which is an EFM violation (or a dimensions spec violation?) so the preparer will receive a message

Comment (chrislococo-wf) [inline dqc_us_rules/dqc_us_0041.py]: Perfect, thank you.

**Comment by chrislococo-wf** [comment]
+1

**Comment by andrewperkins-wf** [comment]
+1

**Comment by chrislococo-wf** [comment]
+1

**Comment by andrewperkins-wf** [comment]
+1

**Comment by chrislococo-wf** [comment]
#### QA Steps: 

- Manual testing: Tested two documents, one that fires, one that originally broke the use case, both pass without issue now and correctly fire 0041 when meant to do so.
- Flake8 on dqc_us_rules: 
```
(dqc)  chrislococo [dqc] (xs-2888-master) $ flake8 dqc_us_rules
(dqc)  chrislococo [dqc] (xs-2888-master) $
```
- Flake8 on tests: 
```
(dqc)  chrislococo [dqc] (xs-2888-master) $ flake8 tests/
(dqc)  chrislococo [dqc] (xs-2888-master) $
```
- Nosetest result: 
```
(dqc)  chrislococo [dqc] (xs-2888-master) $ nosetests tests/
......................................................................
----------------------------------------------------------------------
Ran 70 tests in 0.076s

OK
```

#### Result: QA +10

**Comment by andrewperkins-wf** [comment]
All reviews and tests run.  Merging.
